File: src/main/java/bad/robot/radiate/monitor/ThreadSafeObservable.java
package bad.robot.radiate.monitor;

import bad.robot.radiate.Activity;
import bad.robot.radiate.Progress;
import bad.robot.radiate.Status;

import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.stream.Stream;

import static java.util.Arrays.asList;

public class ThreadSafeObservable implements Observable {

    private final List<Observer> observers = new CopyOnWriteArrayList<>();

    @Override
    public boolean addObservers(Observer... observers) {
        return this.observers.addAll(asList(observers));
    }

    @Override
    public void addObservers(Stream<Observer> observers) {
        observers.forEach(this.observers::add);
    }

    @Override
    public boolean removeObservers(Observer... observers) {
        return this.observers.removeAll(asList(observers));
    }

    @Override
    public void removeAllObservers() {
        observers.clear();
    }

    @Override
    public void notifyObservers(Status status) {
        observers.forEach(observer -> observer.update(this, status));
    }

    @Override
    public void notifyObservers(Exception exception) {
        observers.forEach(observer -> observer.update(this, exception));
    }

    @Override
    public void notifyObservers(Information information) {
        observers.forEach(observer -> observer.update(this, information));
    }

    @Override
    public void notifyObservers(Activity activity, Progress progress) {
        observers.forEach(observer -> observer.update(this, activity, progress));
    }
}

File: src/main/scala/bad/robot/radiate/monitor/ThreadSafeObservable.scala
package bad.robot.radiate.monitor

import java.util.Arrays.asList
import java.util.concurrent.CopyOnWriteArrayList

import bad.robot.radiate.FunctionInterfaceOps._
import bad.robot.radiate.{Activity, Progress, Status}
import bad.robot.radiate.Error

class ThreadSafeObservable extends Observable {

  private val observers = new CopyOnWriteArrayList[Observer]

  def addObservers(observers: Observer*): Boolean = {
    this.observers.addAll(asList(observers:_*))
  }

  def addObservers(observers: List[Observer]) {
    observers.foreach(this.observers.add)
  }

  def removeObservers(observers: Observer*): Boolean = {
    this.observers.removeAll(asList(observers))
  }

  def removeAllObservers() {
    observers.clear()
  }

  def notifyObservers(status: Status) {
    observers.forEach((observer: Observer) => observer.update(this, status))
  }

  def notifyObservers(exception: Exception) {
    observers.forEach((observer: Observer) =>  observer.update(this, exception))
  }

  def notifyObservers(error: Error) {
    observers.forEach((observer: Observer) =>  observer.update(this, error))
  }

  def notifyObservers(information: Information) {
    observers.forEach((observer: Observer) =>  observer.update(this, information))
  }

  def notifyObservers(activity: Activity, progress: Progress) {
    observers.forEach((observer: Observer) =>  observer.update(this, activity, progress))
  }
}